### ISSUE TITLE
In isolated mode, RetryContext.LastFailure.IsCausedBy is not useful if activity function is async

### ISSUE THREAD

**Issue by stevendarby** [Azure/azure-functions-durable-extension#2762]
### Discussed in https://github.com/Azure/azure-functions-durable-extension/discussions/2614

<div type='discussions-op-text'>

<sup>Originally posted by **buckprime** October  3, 2023</sup>
It seems like we need more information about the last error in retry handlers (meaning RetryHandler/AsyncRetryHandler, that you can put into a TaskOptions to specify custom handling of retries). The retry handler gets called with a RetryContext, which has a call LastFailure.IsCausedBy<some exception type>() that is supposed to let you check what exception caused the failure you're retrying. But if the activity function that failed was async, the exception that the RetryContext sees is always AggregateException rather than the actual exception that was thrown. The context only contains the name of the type of the exception, so there's no way to determine what the actual exception was.

I think that the context either needs to know what the original exception was, or provide more information about the exception so that, for instance, if it's an AggregateException we can tell what exceptions it contained. Otherwise IsCausedBy isn't useful if your activity function is async.</div>

**Comment by stevendarby** [comment]
On a related note, is there a way to have a custom handler (`TaskOptions.FromRetryHandler`) & also specify the interval characteristics like you can with `TaskOptions.FromRetryPolicy`? Can `await Task.Delay()` be used in a handler to manually control the intervals, or is that a bad idea?

**Comment by MarioMajcicaAtABNAMRO** [comment]
Indeed, we are having the same issue here. We are moving from inProc to Isolated and aside from having the issue with TaskOptions itself (https://github.com/Azure/azure-functions-durable-extension/issues/2908) we do see that the exception that triggered a retry is not exposed in the handler.

In past we could do something like
`if (ex.InnerException is NotFinishedStatusException statusException && statusException.Operation.Completed)`

Now, we can only check the exception type via
`retryContext.LastFailure.IsCausedBy<NotFinishedStatusException>()`

It would be handy if the exception itself is exposed.

**Comment by thomaseyde** [comment]
I ran into the same issue today. I would think that activity functions would mostly be async, otherwise there is very little useful work they can do.

I have looked into how `IsCausedBy` works:

The owner record `TaskFailureDetails` is recursive, it has a property `InnerFailure`. There is a placeholder to keep inner exception details, but for some reason it is not used. The debugger also tells me this value is `null`, so I cannot create my own workaround.

A higher priority than P3 is desired.

**Comment by thomaseyde** [comment]
~~`IsCausedBy` seems to have been removed from `RetryContext` in `1.1.7`.~~

My bad. Was mistakenly looking at the root object, not the property.

**Comment by cgillum** [comment]
@thomaseyde this may have been mentioned in another GitHub issue, but `IsCausedBy` is defined on the `RetryContext.LastFailure` instance (which is of type `TaskFailureDetails`). Looking at the GitHub history, I don't think any SDK changes were made in the `1.1.7` release, and I don't think the `RetryContext` class has ever changed.

**Comment by cgillum** [comment]
I'm raising the priority of this item to P1. Showing `AggregateException` on async activities is not the intended behavior, and I agree that it's not particularly useful.

/cc @lilyjma

**Comment by codywilliamson** [comment]
Has there been any updates regarding this issue? We are using the latest SDK and are being plagued by this. We just moved to isolated and are having to rework how we handle exceptions in activities across the board atm since we don't have any way to capture the original exception.

**Comment by nytian** [comment]
@stevendarby  Do you have a repro where IsCausedBy returns an AggregateException, and which version are you using? I tested this with version 1.1.7, where IsCausedBy returns `unknown` exception type and this is a known bug that has been fixed. And with the latest versionv1.3.0 , it correctly returns the specific exception type.

**Comment by nytian** [comment]
@codywilliamson Thanks for the feedback! Could you provide a minimal repro for this issue? Also, if you're referring to the original exception, we've introduced some fixes for it in the latest release. Please try upgrading to version [v1.3.0](https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker.Extensions.DurableTask/1.3.0), which includes those improvements #3034 and #3070. You should be able to access the original exception via TaskFailedException.FailureDetails.InnerException, etc.

**Comment by codywilliamson** [comment]
@nytian 

I ended up going through MS support within my company and worked with the PG to reach a solution. We updated to the latest update and while it doesn't give us the same functionality we had originally, we are now seeing a properly populated FailureDetails object. Thanks.

**Comment by mmajcica** [comment]
> [@codywilliamson](https://github.com/codywilliamson) Thanks for the feedback! Could you provide a minimal repro for this issue? Also, if you're referring to the original exception, we've introduced some fixes for it in the latest release. Please try upgrading to version [v1.3.0](https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker.Extensions.DurableTask/1.3.0), which includes those improvements [#3034](https://github.com/Azure/azure-functions-durable-extension/pull/3034) and [#3070](https://github.com/Azure/azure-functions-durable-extension/pull/3070). You should be able to access the original exception via TaskFailedException.FailureDetails.InnerException, etc.

Still there is no way to get the exception itself. Which is unfortunate as in the In-Proc version we based our retry decision not only on the exception type, but also on some data that we had in the exception. This is making the move to isolated worker more difficult.
Has anything changed or am I missing something? Is there a way to get the exception itself and not just the TaskFailureDetails?

**Comment by nytian** [comment]
@mmajcica  Thanks for the information.

Sorry for the inconvenience — using FailureDetails is part of the design and cannot be changed, as it helps prevent serialization issues.
To access exception data or custom properties, we’ve recently added support for this. Please see the documentation here: 
https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-error-handling?tabs=csharp-isolated#include-custom-exception-properties-for-failuredetails-net-isolated

**Comment by nytian** [comment]
Close as the issue is solved. Please check document here for error handling details. Also please let me know if there is any issue or problem. Thanks! https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-error-handling?tabs=csharp-isolated
